### ISSUE TITLE
Implement refresh tokens in ArgoCD Web UI

### ISSUE THREAD

**Issue by jkroepke** [argoproj/argo-cd#8325]
Hi, I'm coming from https://github.com/argoproj/argo-cd/issues/455

# Summary

Implement refresh tokens in ArgoCD Web UI.

# Motivation

For security reasons, all access tokens are short-living (5 minutes) and refresh token (12 hours).

I full understand the "re-logging in should be as convenient as possible" these, but I opened an application in ArgoCD Web UI. It wont refresh the UI. After 5 minutes, I only see a lot of 401 requests. I dont get any notification that my session is expired and I need to refresh the page manually. This is confuse a lot of peoples.

# Proposal

## 1. Implement refresh tokens

If a token is expired, use the refresh token to refresh an existing token. As I know, the refresh token can be stored inside browser as cookie, too.

## 2. Do a full browser reload, if sessions is expired.

Is may not fully fix the problem, but it helps a lot since the Web UI can't be out of date anymore. A full browser reload is fine here, because from IDP perspective, I'm still logged in and get instantly redirected back to ArgoCD.

**Comment by jkroepke** [comment]
To gain more information, why we have such a low access token lifetime in our system:

https://www.pcidssguide.com/pci-dss-session-timeout-requirements/

> PCI DSS requirement 8.1.8 requires the user to re-authenticate to reactivate the terminal or session if a session has been idle for more than 15 minutes. The PCI DSS inactive session timeout requirement applies to administrative or internal accounts.

From my point of view, solutions like https://github.com/argoproj/argo-workflows/pull/4095 won't help here, since the JWE generated by ArgoCD must have a short life access time, otherwise it might be a conflict with PCI.

**Comment by MrLuje** [comment]
@jkroepke I understand the point for idle session but what about refreshing tokens' on explicit user actions ? (clicking "Refresh", "Sync" buttons, etc)

**Comment by jkroepke** [comment]
Good point, explicit user action should be supported as well as idle sessions

**Comment by ricardojdsilva87** [comment]
Hello @jkroepke , thanks for this initiative!
I have added some information that happens to us [in this ticket](https://github.com/argoproj/argo-cd/issues/12189#issuecomment-1462141620)
It seems that it's happening exactly as you describe

**Comment by Hronom** [comment]
Same story here, even with 1 hour token it's very disruptive. Please can you do this re-login in background?

**Comment by GeorgiYosifov** [comment]
Can I work on this after review of this PR: https://github.com/argoproj/argo-cd/pull/15004

**Comment by michaelajr** [comment]
This issues has 50 upvotes. It is important. OIDC access/ID tokens are short lived - 5m sometimes. The refresh_token grant is used to implicitly get new tokens without the user needing to fully re-authenticate every 5m. The grant also returns a new refresh token on each call but with its expiration ticked down. When it reaches the total lifetime - that's when users should re-auth.

**Comment by mpsOxygen** [comment]
This would be really useful for us as well. Ended up here while searching how to enable refresh tokens in ArgoCD.

**Comment by vfauth** [comment]
It is a requirement in my company too. Short-lived OIDC token are a security requirement but login every 5 minutes can be quite frustrating.

**Comment by askerrit** [comment]
Hi, I am willing to contribute dev hours to this

I commented this back in June:
https://github.com/argoproj/argo-cd/pull/15004#issuecomment-2144605427

What can we do to help? We are all developers who need this. I am sure we can contribute with some guidance :)

**Comment by andrii-korotkov-verkada** [comment]
Consolidating with https://github.com/argoproj/argo-cd/issues/12189. @devopsjedi is working on the refresh token implementation.
